rogg_granule.c

```diff
@@ -0,0 +1,162 @@
+/*
+   Copyright (C) 2007 Xiph.org Foundation
+   Copyright (C) 2009,2016 Joern 'manx' Heusipp
+
+   Redistribution and use in source and binary forms, with or without
+   modification, are permitted provided that the following conditions
+   are met:
+
+   - Redistributions of source code must retain the above copyright
+   notice, this list of conditions and the following disclaimer.
+
+   - Redistributions in binary form must reproduce the above copyright
+   notice, this list of conditions and the following disclaimer in the
+   documentation and/or other materials provided with the distribution.
+
+   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+   ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+   PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE ORGANISATION OR
+   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+   EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+   PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+   PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+   LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+/* ogg vorbis logical stream granule adjustment using the rogg library */
+
+/* compile with
+   gcc -O2 -g -Wall -I. -o rogg_granule rogg.c rogg_granule.c
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include <rogg.h>
+
+int granule_adjust = 0;
+
+void print_usage(FILE *out, char *name)
+{
+  fprintf(stderr, "Script for editing ogg headers\n");
+  fprintf(stderr, "%s [-g n] <file1.ogg> [<file2.ogg>...]\n",
+	name);
+  fprintf(stderr,
+		  "    -g n        change the granule position of a logical vorbis stream by n\n"
+		  "                (can be positive or negative)\n"
+		  "\n");
+}
+
+int parse_args(int *argc, char *argv[])
+{
+  int arg = 1;
+  int shift;
+
+  while (arg < *argc) {
+    shift = 0;
+    if (argv[arg][0] == '-') {
+      switch (argv[arg][1]) {
+	case 'g':
+	  /* read granule adjustment from the next arg */
+	  if (sscanf(argv[arg+1], "%i", &granule_adjust) != 1) {
+	    fprintf(stderr, "Could not parse granule adjustment '%s'.\n", argv[arg+1]);
+	    granule_adjust = 0;
+	  } else {
+	    fprintf(stdout, "Adjusting granule position by %i\n", granule_adjust);
+	  }
+	  shift = 2;
+	  break;
+      }
+    }
+    if (shift) {
+      int left = *argc - arg - shift;
+      if (left < 0) {
```

Comment:
I see, argc is still trimmed by shift even if there are no remaining arguments to move down. Ok as-is.